pkg/controllers/grafana-operator/controller.go

```diff
@@ -151,108 +158,161 @@ func RegisterWithManager(mgr controllerruntime.Manager) error {
+	if res := r.reconcileNamespace(ctx); res.stop {
+		return result(res)
 	}
 
-	r.logger.V(10).Info("Reconciling Grafana Operator Subscription")
-	if err := r.reconcileSubscription(ctx); err != nil {
-		return controllerruntime.Result{}, err
+	logger.Info("Reconciling Grafana Operator OperatorGroup")
+	if res := r.reconcileOperatorGroup(ctx); res.stop {
+		return result(res)
 	}
 
-	r.logger.V(10).Info("Reconciling Grafana instance")
-	if err := r.reconcileGrafana(ctx); err != nil {
-		return controllerruntime.Result{}, err
+	logger.Info("Reconciling Grafana Operator Subscription")
+	if res := r.reconcileSubscription(ctx); res.stop {
+		return result(res)
 	}
 
-	return controllerruntime.Result{}, nil
+	logger.Info("Reconciling Grafana instance")
+	if res := r.reconcileGrafana(ctx); res.stop {
+		return result(res)
+	}
+
+	return ctrl.Result{}, nil
 }
 
-func (r *reconciler) reconcileNamespace(ctx context.Context) error {
-	key := types.NamespacedName{
-		Name: Namespace,
-	}
+func (r *reconciler) reconcileNamespace(ctx context.Context) reconcileResult {
+
+	key := types.NamespacedName{Name: Namespace}
 	var namespace corev1.Namespace
 	err := r.k8sClient.Get(ctx, key, &namespace)
-	if err == nil {
-		r.logger.V(10).Info("Namespace already exists")
-		return nil
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcileError(err)
 	}
 
 	if errors.IsNotFound(err) {
 		r.logger.Info("Creating namespace", "Namespace", namespace)
-		return r.k8sClient.Create(ctx, NewNamespace())
+		err = r.k8sClient.Create(ctx, NewNamespace())
+		return creationResult(0, err)
 	}
 
-	r.logger.Error(err, "error reconciling namespace ")
-	return err
+	///
+	// requeue if namespace is marked for deletion
+	// TODO(sthaha): decide if want to use finalizers to prevent deletion but
+	// we also need to solve how to properly cleanup / uninstall operator
+
+	if namespace.Status.Phase != corev1.NamespaceActive {
+		r.logger.Info("namespace is present but not active", "phase", namespace.Status.Phase)
+		return requeue(5*time.Second, nil)
+	}
+	return next()
 }
 
-func (r *reconciler) reconcileOperatorGroup(ctx context.Context) error {
+func (r *reconciler) reconcileOperatorGroup(ctx context.Context) reconcileResult {
 	key := types.NamespacedName{
 		Name:      operatorGroupName,
 		Namespace: Namespace,
 	}
 	var operatorGroup operatorsv1.OperatorGroup
+
 	err := r.k8sClient.Get(ctx, key, &operatorGroup)
-	if err == nil {
-		r.logger.Info("Updating Grafana Operator OperatorGroup")
-		operatorGroup.Spec = NewOperatorGroup().Spec
-		return r.k8sClient.Update(ctx, &operatorGroup)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconcileError(err)
 	}
+
+	// create
 	if errors.IsNotFound(err) {
-		r.logger.Info("Creating Grafana Operator OperatorGroup")
-		return r.k8sClient.Create(ctx, NewOperatorGroup())
+		r.logger.Info("Creating Grafana OperatorGroup")
+		err := r.k8sClient.Create(ctx, NewOperatorGroup())
+		return creationResult(5*time.Second, err)
```

Comment:
Do we need an explicit requeue here? Since there's a watcher on the resource, it will trigger a new reconcile again after creation.

Reply:
You are right, I don't think we should need to requeue! Good catch!

Reply:
I have updated the `creationResult` logic to terminate the loop if there are not creation errors